docs/internals/layout_in_memory.rst

```diff
@@ -1,40 +1,113 @@
+always places new objects at the to location pointed to by the free memory pointer. Memory is never
+"freed" (this might change in the future). In other words, in the current version of Solidity, the
+value at ``0x40`` is never reduced by Solidity.
+
+.. warning::
+  There are some operations in Solidity that need a temporary memory area
+  larger than 64 bytes and therefore will not fit into the scratch space.
+  Code for these operations therefore happens at the location pointed to
+  by the free memory pointer. Because such operations are known to be complete
+  before the next instruction or operation managed by Solidity, the generated
+  code does not update the free memory pointer. The memory used in such operations
+  may or may not be zeroed out after the operation is complete. Solidity's
+  access of that memory may affect ``msize()``. Thus, it's reasonable to expect that
+  ``msize()`` might be larger than the value of the free memory pointer.
+
+  While it may seem like a good idea to use ``msize()`` to arrive at a
+  definitely zeroed out memory area (by using the memory just beyond its
+  return value) using such a pointer non-temporarily without updating the
+  free memory pointer can have unexpected results. Nothing prevents inline aseembly
+  from accessing any memory location whatsoever, but updating the free memory pointer
+  at ``0x40`` at least ensures that Solidity iself won't unexpectedly use memory
+  located below its value.
+  
+  In brief, the memory between ``0x00 - 0x3F`` and the memory after the location
+  pointed to by the free memory pointer can be considered scratch memory, and should
+  not be used except between Solidity instructions via inline assembly, if
+  you know what you are doing.
+
 
-Solidity always places new objects at the free memory pointer and
-memory is never freed (this might change in the future).
+****************
+Layout in Memory
+****************
+
+Layout of Arrays In Memory
+================================
 
-Elements in memory arrays in Solidity always occupy multiples of 32 bytes (this
-is even true for ``bytes1[]``, but not for ``bytes`` and ``string``).
+Elements in memory arrays in Solidity always occupy multiples of 32 bytes, i.e.
+one slot (this is even true for ``bytes1[]``, but not for ``bytes`` and ``string``).
 Multi-dimensional memory arrays are pointers to memory arrays. The length of a
 dynamic array is stored at the first slot of the array and followed by the array
 elements.
 
+Dynamically sized memory arrays (e.g., `uint[] varSizedArray`) store their length in
+the first slot pointed to by the array variable, followed by its elements in the following
+slots. Although memory arrays are not resizable, they may be allocated dynamically using
+the `new` keyword. The initial value of the dynamically sized memory array variable itself,
+is ``0x60``, see :ref:`Reserved Memory Areas and Memory Management Details`, a slot that
```

Comment:
This is also a bit of an implementation detail/optimization. I don't think we guarantee that it always is (and will be) the case.
```suggestion
the `new` keyword. An empty dynamically sized memory array variable may point at the zero slot, see :ref:`Reserved Memory Areas and Memory Management Details`, a slot that
```